### ISSUE TITLE
Does this work well with Rails 5.2+?

### ISSUE THREAD

**Issue by 9mm** [heroku/heroku-buildpack-pgbouncer#121]
I wanted to use PGBouncer in transaction mode in my Rails 5.2 app. I am aware that it breaks session-level features in my own code, such as advisory locks.

One thing I wasn't aware of though, is that it appears it can break Rails itself...

https://github.com/rails/rails/issues/22102
https://electron0zero.xyz/blog/til-connection-pooling-and-pgbouncer

I can't tell if this is older information though. Are these problems still prevalent with the latest Rails do you know?

**Comment by MrHubble** [comment]
@9mm were you able to find any more information on this?

**Comment by 9mm** [comment]
@MrHubble No I never found any more information on it, which deterred me from using it for now. I'd love to know if you discover anything more :/

**Comment by kevine-voicera** [comment]
We have been struggling to get this to work well too.

**Comment by 9mm** [comment]
@kevine-voicera what part specifically is giving you trouble? Is it default vanilla rails working with PG Bouncer?

**Comment by kevine-voicera** [comment]
We had issues with vanilla Rails 5.2 migrations from executing properly due to disabled statements / advisory locks.

**Comment by mble** [comment]
Hey @kevine-voicera, there is a documented workaround regarding how Rails 5.2+ handles migrations with advisory locks and PGBouncer here: https://help.heroku.com/UYH8N2WW/why-do-i-receive-activerecord-concurrentmigrationerror-when-running-rails-migrations-using-pgbouncer

Typically, you'll want to patch AR (as per Discourse) with something like: https://github.com/discourse/discourse/blob/master/lib/freedom_patches/postgresql_adapter.rb

```ruby
# Awaiting decision on https://github.com/rails/rails/issues/31190
if ENV['DISABLE_MIGRATION_ADVISORY_LOCK']
  class ActiveRecord::ConnectionAdapters::PostgreSQLAdapter
    def supports_advisory_locks?
      false
    end
  end
end
```

Alternatively, in Rails 6.0+ https://github.com/rails/rails/pull/33691 should be available, enabling configuration with patching like this.

**Comment by 9mm** [comment]
@mble thats amazing. is that the only thing in rails that is incompatible with PGBouncer? That's really awesome news

**Comment by mble** [comment]
@9mm It's very dependent on your use cases and the queries you are running – it's entirely possible to end up needing some of the session-level features that are broken in the `transaction` pooling mode.

It's possible there are some other incompatibilities that I'm not aware of, but haven't seen any major incompatibilities outside of the ones raised above and elsewhere in issues in this repo.

**Comment by 9mm** [comment]
Yeah sorry I should have made that clear. I mean just with vanilla rails compatibility + pgbouncer.

For heroku I wouldn't run pgbouncer on the worker/scheduler dynos, only puma dyno + rails 5.2. No queries in the app use advisory locks outside of jobs.

**Comment by mble** [comment]
No worries! Sure, I haven't seen much in the way of incompatibility in that setup, though as always it's worth monitoring and testing out those kinds of changes in a staging environment or shadow production.

**Comment by kevine-voicera** [comment]
@mble This worked for us. Of course it means we need to deploy multiple production apps in the pipeline independently if there are migrations (to ensure they don't stomp on each other).
